### ISSUE TITLE
Use Monaco to replace CodeMirror

### ISSUE THREAD

**Issue by Unknwon** [gogits/gogs#4201]
[Monaco](https://github.com/Microsoft/monaco-editor) editor seems promising and very nice. By no means to make a IDE for Gogs, but can be a better alternative for Gogs for online editing.

**Comment by stefan-lacatus** [comment]
Yes, i've had some recent experience with replacing codemirror with monaco in a project and  it was a quite easy, while providing impressive results. 
I'll try to provide a PR but, it may take 1-2 weeks since I have zero experience with GO and the gogs source code.

**Comment by stefan-lacatus** [comment]
I've investigated the source code a bit and don't think it would be a very complicated change to do (mostly changes in gogs.js)
A couple of questions:
- The monaco editor code should be included under `public/plugins/monaco-0.8.1`?
- Should we do a complete switch to Monaco, or is there supposed to be a user preference selecting between CodeMirror and Monaco?

**Comment by Unknwon** [comment]
Just keep in mind, it is not simple editor file replacement, you need coding highlighting, markdown editor. From what I saw Monaco is not as mature as CodeMirror yet, its high level encapsulations are very few.

**Comment by Unknwon** [comment]
If we can't fully switch to Monaco, it's not worth adding it at all. 😣

**Comment by stefan-lacatus** [comment]
Code editor and syntax highlighting would be easy to add. 
Monaco has very good support for javascript, typescript, json, html and css. This includes advanced code completion, snippets other language services, for example "go to definition".
Then, for all the languages [here](https://github.com/Microsoft/monaco-languages), monaco only has colorization support.

However, editing Markdown is nowhere near as advanced as what already exists in SimpleMDE. Is this a good enough trade off?

**Comment by Unknwon** [comment]
>However, editing Markdown is nowhere near as advanced as what already exists in SimpleMDE. Is this a good enough trade off?

This is actually the biggest drawback for me. Can't understand why there is not even a Markdown editor for Monaco right now. Maybe I'm missing something...
